File: bench/stress_hugind.py
import json
import time
import argparse
import http.client
import threading
import statistics
import random
from urllib.parse import urlparse

# Configuration for Retries
MAX_RETRIES = 5
BACKOFF_FACTOR = 2  # Seconds to wait between retries (multiplied by attempt number)

def benchmark_request(base_url, model, prompt, request_id, results_list):
    parsed_url = urlparse(base_url)
    host = parsed_url.hostname
    port = parsed_url.port or (443 if parsed_url.scheme == 'https' else 80)
    path = parsed_url.path.rstrip('/') + "/chat/completions"

    payload = {
        "model": model,
        "messages": [{"role": "user", "content": prompt}],
        "stream": True,
        "max_tokens": 100
    }
    
    headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer nopass",
    }

    attempt = 0
    while attempt < MAX_RETRIES:
        start_time = time.time()
        ttft = None
        token_count = 0
        
        try:
            conn = http.client.HTTPConnection(host, port, timeout=30)
            conn.request("POST", path, body=json.dumps(payload), headers=headers)
            response = conn.getresponse()
            
            # Check if server rejected request (e.g., 429 Too Many Requests or 503)
            if response.status != 200:
                raise Exception(f"HTTP {response.status}")

            while True:
                line = response.fp.readline()
                if not line:
                    break
                    
                line = line.decode('utf-8').strip()
                if not line or not line.startswith("data: "):
                    continue
                    
                data_str = line[6:]
                if data_str == "[DONE]":
                    break
                    
                try:
                    data = json.loads(data_str)
                    if not ttft:
                        ttft = time.time() - start_time
                    
                    if "choices" in data and len(data["choices"]) > 0:
                        delta = data["choices"][0].get("delta", {})
                        if "content" in delta:
                            token_count += 1
                except json.JSONDecodeError:
                    continue

            end_time = time.time()
            conn.close()

            total_time = end_time - start_time
            tpot = (total_time - ttft) / (token_count - 1) if token_count > 1 else 0

            results_list.append({
                "request_id": request_id,
                "ttft": ttft,
                "total_time": total_time,
                "token_count": token_count,
                "tpot": tpot
            })
            
            # REQUIREMENT: Print when finished
            print(f"req {request_id} finished in {total_time:.2f} seconds")
            return # Success! Exit the retry loop

        except Exception as e:
            attempt += 1
            wait_time = (BACKOFF_FACTOR * attempt) + random.uniform(0, 1)
            if attempt < MAX_RETRIES:
                # print(f"req {request_id} failed ({e}), retrying in {wait_time:.1f}s...")
                time.sleep(wait_time)
            else:
                print(f"req {request_id} failed permanently after {MAX_RETRIES} attempts.")

def main():
    parser = argparse.ArgumentParser(description="LLM Benchmark with Retries")
    parser.add_argument("--base-url", type=str, default="http://localhost:8080/v1")
    parser.add_argument("--model", type=str, default="gpt-3.5-turbo")
    parser.add_argument("--total", type=int, default=100, help="Total requests to send")
    parser.add_argument("--concurrency", type=int, default=25, help="Max parallel requests")
    parser.add_argument("--prompt", type=str, default="Explain quantum physics in one sentence.")
    
    args = parser.parse_args()

    print(f"Sending {args.total} total requests (Concurrency: {args.concurrency})")
    print("-" * 50)

    threads = []
    results = []
    
    # Use a Semaphore to limit active threads to your hardware capacity (25)
    # but still loop until we reach the total (100)
    semaphore = threading.Semaphore(args.concurrency)

    def worker(req_id):
        with semaphore:
            benchmark_request(args.base_url, args.model, args.prompt, req_id, results)

    for i in range(args.total):
        t = threading.Thread(target=worker, args=(i,))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    # Final Statistics
    if not results:
        return

    total_times = [r["total_time"] for r in results]
    print("-" * 50)
    print(f"Completed {len(results)}/{args.total} requests successfully.")
    print(f"Average response time: {statistics.mean(total_times):.2f}s")

if __name__ == "__main__":
    main()

File: bench/stress_hugind.dart
import 'dart:convert';
import 'dart:io';

import 'package:args/args.dart';
import 'package:http/http.dart' as http;

class BenchResult {
  BenchResult({
    required this.requestId,
    required this.ttft,
    required this.totalTime,
    required this.tokenCount,
    required this.tpot,
  });

  final int requestId;
  final double? ttft;
  final double totalTime;
  final int tokenCount;
  final double tpot;
}

Future<BenchResult?> benchmarkRequest({
  required http.Client client,
  required String baseUrl,
  required String model,
  required String prompt,
  required int requestId,
  required Map<String, int> stats,
}) async {
  final baseUri = Uri.parse(baseUrl);
  final path = baseUri.path.replaceAll(RegExp(r'/+$'), '') + '/chat/completions';
  final requestUri = baseUri.replace(path: path);

  final payload = <String, dynamic>{
    'model': model,
    'messages': [
      {'role': 'user', 'content': prompt}
    ],
    'stream': true,
    'max_tokens': 100,
  };

  final request = http.Request('POST', requestUri);
  request.headers.addAll({
    'Content-Type': 'application/json',
    'Authorization': 'Bearer nopass',
  });
  request.body = jsonEncode(payload);

  final stopwatch = Stopwatch()..start();
  double? ttft;
  var tokenCount = 0;

  try {
    final response = await client
        .send(request)
        .timeout(const Duration(seconds: 300));
    if (response.statusCode != 200) {
      await response.stream.drain();
      if (response.statusCode == 429) {
        stats['queued'] = (stats['queued'] ?? 0) + 1;
      } else if (response.statusCode == 503) {
        stats['rejected'] = (stats['rejected'] ?? 0) + 1;
      } else {
        stats['errors'] = (stats['errors'] ?? 0) + 1;
      }
      throw HttpException('HTTP ${response.statusCode}');
    }

    await for (final line in response.stream
        .transform(utf8.decoder)
        .transform(const LineSplitter())) {
      final trimmed = line.trim();
      if (trimmed.isEmpty || !trimmed.startsWith('data: ')) {
        continue;
      }

      final dataStr = trimmed.substring(6);
      if (dataStr == '[DONE]') {
        break;
      }

      try {
        final data = jsonDecode(dataStr) as Map<String, dynamic>;
        ttft ??= stopwatch.elapsedMicroseconds / 1e6;
        final choices = data['choices'];
        if (choices is List && choices.isNotEmpty) {
          final delta = choices.first['delta'];
          if (delta is Map<String, dynamic> && delta['content'] != null) {
            tokenCount += 1;
          }
        }
      } on FormatException {
        continue;
      }
    }
  } catch (e) {
    stdout.writeln('req $requestId failed: $e');
    return null;
  } finally {
    stopwatch.stop();
  }

  final totalTime = stopwatch.elapsedMicroseconds / 1e6;
  final tpot = (ttft != null && tokenCount > 1)
      ? (totalTime - ttft) / (tokenCount - 1)
      : 0.0;

  stats['success'] = (stats['success'] ?? 0) + 1;
  stdout.writeln('req $requestId finished in ${totalTime.toStringAsFixed(2)} seconds');

  return BenchResult(
    requestId: requestId,
    ttft: ttft,
    totalTime: totalTime,
    tokenCount: tokenCount,
    tpot: tpot,
  );
}

double mean(List<double> values) =>
    values.isEmpty ? 0.0 : values.reduce((a, b) => a + b) / values.length;

Future<void> main(List<String> args) async {
  final parser = ArgParser()
    ..addOption('base-url', defaultsTo: 'http://localhost:8080/v1')
    ..addOption('model', defaultsTo: 'gpt-3.5-turbo')
    ..addOption('total', defaultsTo: '100', help: 'Total requests to send')
    ..addOption('prompt',
        defaultsTo: 'Explain quantum physics in one sentence.');

  final parsed = parser.parse(args);
  final baseUrl = parsed['base-url'] as String;
  final model = parsed['model'] as String;
  final total = int.tryParse(parsed['total'] as String) ?? 100;
  final prompt = parsed['prompt'] as String;

  stdout.writeln('Sending $total total requests');
  stdout.writeln('-' * 50);

  final stats = {'success': 0, 'queued': 0, 'rejected': 0, 'errors': 0};
  final client = http.Client();
  List<BenchResult> results;
  try {
    final futures = List.generate(
      total,
      (i) => benchmarkRequest(
        client: client,
        baseUrl: baseUrl,
        model: model,
        prompt: prompt,
        requestId: i,
        stats: stats,
      ),
    );
    results = (await Future.wait(futures)).whereType<BenchResult>().toList();
  } finally {
    client.close();
  }

  if (results.isEmpty) {
    return;
  }

  final totalTimes = results.map((r) => r.totalTime).toList();
  stdout.writeln('-' * 50);
  stdout.writeln('Completed ${results.length}/$total requests successfully.');
  stdout.writeln('Queued (429): ${stats['queued']}');
  stdout.writeln('Rejected (503): ${stats['rejected']}');
  stdout.writeln('Other errors: ${stats['errors']}');
  stdout.writeln(
      'Average response time: ${mean(totalTimes).toStringAsFixed(2)}s');
}
